Virtual values

fickludd opened the pull request:
Adds virtual values to the values module. These are

 * NodeReference
 * EdgeReference
 * LabelSet
 * Path
 * Point (cartesian and geographic)
 * List
 * Map

LabelSet is included so that full nodes can be handled (as the combination of 1 NodeReference, 1 LabelSet, and 1 Map). 

These virtual values support `equality`, `hashcode`, `comparison` and `writeTo`. Two known changes are not performed here. We have created card for completing these work units at some later point. The left out changes are

 1) to flip the packages to the storable values are in a self contained subpackage, while the virtual/AnyValues are in the top values package.
 2) implement list-array equality

fickludd committed:
First structure for virtual values + ListValue and MapValue

fickludd committed:
Test bug fix and test util refactoring

fickludd committed:
LabelSet as value

I'm a bit sceptical against this solution, as a LabelSet cannot be
handled as a value in Cypher. This is just a trick to avoid having
fat node values.

fickludd committed:
Fat node and edge value

Will refactor into reference in the commit. Just checked these in so
we can revive them if we need them later.

fickludd committed:
Reintroduce NodeValue and NodeRelationship

Having only id makes it unnecessarily hard to implement writers.

fickludd committed:
NodeReference, EdgeReference and PathValue

fickludd committed:
PointValue

fickludd committed:
WriteTo tests and implementation for all virtual values

fickludd committed:
Virtual value comparison

pontusmelke commented on community/values/src/main/java/org/neo4j/values/virtual/ArrayHelpers.java:
```diff
@@ -0,0 +1,55 @@
+ *
+ * Should we introduce dependency on primitive collections?
+ */
+class ArrayHelpers
```
we usually have static helpers being final with a `private` constructor so that they cannot be instantiated

pontusmelke commented on community/values/src/main/java/org/neo4j/values/virtual/EdgeReference.java:
```diff
@@ -0,0 +1,85 @@
+    @Override
+    public boolean equals( VirtualValue other )
+    {
+        if ( other == null || !(other instanceof EdgeReference) )
```
isn't this null check redundant?

pontusmelke commented on community/values/src/main/java/org/neo4j/values/virtual/MapValue.java:
```diff
@@ -0,0 +1,160 @@
+    @Override
+    public boolean equals( VirtualValue other )
+    {
+        if ( other == null || !(other instanceof MapValue) )
```
same here, redundant?

pontusmelke commented on community/values/src/main/java/org/neo4j/values/virtual/NodeReference.java:
```diff
@@ -0,0 +1,85 @@
+    @Override
+    public boolean equals( VirtualValue other )
+    {
+        if ( other == null || !(other instanceof NodeReference) )
```
Same here, redundant?

pontusmelke commented on community/values/src/main/java/org/neo4j/values/virtual/PathValue.java:
```diff
@@ -0,0 +1,136 @@
+    @Override
+    public boolean equals( VirtualValue other )
+    {
+        if ( other == null || !(other instanceof PathValue) )
```
same here, redundant?

pontusmelke committed:
Make ValueWriter and writeTo throw exceptions

In many cases the writer will write to disk or network buffers where
most thing will throw IOExceptions. We should allow a `ValueWriter` and
 `Value.writeTo` to throw along these exceptions and not force
 implementations to catch and rethrow.

fickludd committed:
Made static classes final w. private constructors

fickludd committed:
Fixed instanceof checks in virtual values

pontusmelke committed:
Don't use Optional as field

Optional shouldn't be used as a field.

pontusmelke committed:
Add LabelValue as a value type

Introduce `LabelValue`, and now a `LabelSet` is a collection
of these instead of just an array of ints. Also update ValueWriter
to write a `LabelValue` instead of just obtaining the label token.

pontusmelke committed:
Add writeNode method to ValueWriter

When writing a fat node we want the writer to be provided with both
labels and properties as well as node ids.

pontusmelke committed:
Changed so that MapValue is backed by HashMap

fickludd commented on community/values/src/main/java/org/neo4j/values/virtual/VirtualValues.java:
```diff
@@ -32,19 +33,32 @@ private VirtualValues()
         return new ListValue( values );
     }
 
+    public static MapValue emptyMap()
```
This can be a static field instead of a method.

fickludd commented on community/values/src/main/java/org/neo4j/values/virtual/VirtualValues.java:
```diff
@@ -32,19 +33,32 @@ private VirtualValues()
+        return new LabelValue( id, value );
+    }
+
+    public static LabelSet labels( LabelValue... labelIds )
```
arguments name could be `labels`

fickludd commented on community/values/src/test/java/org/neo4j/values/AnyValueComparatorTest.java:
```diff
@@ -63,11 +68,13 @@
             node( 1L ),
-            node( 2L ),
+            nodeValue( 2L, labels( label( 1, stringValue( "L" ) ) ), emptyMap() ),
+            node( 3L ),
```
WDYT about `node` -> `nodeRef`, and same for edge?

fickludd commented on community/values/src/main/java/org/neo4j/values/virtual/NodeValue.java:
```diff
@@ -40,12 +42,19 @@
+    @Override
+    public String toString()
+    {
+        return format( "(%s)", id );
```
`"(%d)"`

fickludd reviewed (commented):
Looks good! I will be happy to use some benchmarks later to find a faster MapValue implementation =).

pontusmelke committed:
Fixes from code review

Fixed
 - toString-implementations of Node and Edge value implementations
 - fixed argument name to `labels` method

pontusmelke committed:
Changed to use bolt path encoding

BOLT uses a very particular path encoding, initially the writeTo of
PathValue now completely mimics the bolt encoding but at some point when
there are more users AnyValueWriter there may be reason to move some of
that logic to the implementing class.

pontusmelke committed:
Simplified path serialization

Having the full complicated path serialization living in PathValue
does not make sense. Simplify so that `PathValue.writeTo` just writes
the nodes and edges to the writer and let that figure out the preferred
serialization scheme.

pontusmelke committed:
Add special method for writing byte[]

Add a primitive for writing a byte[] in bulk instead of
just one value at a time.

pontusmelke commented on community/values/src/main/java/org/neo4j/values/virtual/LabelValue.java:
```diff
@@ -0,0 +1,87 @@
+import org.neo4j.values.TextValue;
+import org.neo4j.values.VirtualValue;
+
+public class LabelValue extends VirtualValue
```
I have changed my mind, this was not a good idea

pontusmelke committed:
Remove LabelValue

pontusmelke committed:
Add explicit stringOrNoValue method

pontusmelke committed:
Made MapValue.compareTo more robust

 Old implementation was relying on the order items are returned from
 HashMap

pontusmelke committed:
Made LabelSet.compareTo more robust

chrisvest commented on community/values/src/main/java/org/neo4j/values/Values.java:
```diff
@@ -375,19 +387,19 @@ public static Value of( Object value, boolean allowNull )
-     *
-     * Converts an array of object values to the internal Value type. See {@Values.of}.
+     * <p>
+     * Converts an array of object values to the internal Value type. See @{Values.of}.
```
`{@link Values#of}`?

pontusmelke committed:
Remove LabelSet
